### ISSUE TITLE
[java] UnnecessaryImport false positive for on-demand imports

### ISSUE THREAD

**Issue by pzygielo** [pmd/pmd#2655]
**Affects PMD Version:**
~6.25~ 6.34.0 (this ticket was originally for UnusedImports, replaced by UnnecessaryImport)

**Rule:**
[UnnecessaryImports](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#unnecessaryimport)

**Description:**
Used IMO [star-import reported as](https://github.com/pzrep/pmd-UnusedImports/runs/886787172#step:5:41)
```
[INFO] PMD Failure: m.U:3 Rule:UnusedImports Priority:4 Avoid unused imports such as 'p'.
```

**Code Sample demonstrating the issue:**
[maven project](https://github.com/pzrep/pmd-UnusedImports) with:
```java
package p;

public class C {
  final public static String V = "";
}
```
```java
package m;

import p.*;

public class U {
  private void g() {
    String k = C.V;
  }
}
```

**Expected outcome:**
No violation reported, as `p.C.V` is resolved from `p.*`.

*   Does PMD report a violation, where there shouldn't be one? -> false-positive
Yes it does.

*   Is PMD missing to report a violation, where there should be one? -> false-negative
No.

**Running PMD through:**
maven

Similar open issues, but reported for JavaDoc usage of import, while I observe it with code
* #2304
* #2001
* #1639

Similar and closed
 * #2016 - delivered with 6.20.0
But with 6.20.0 I also experience violation in my case.

It's possible I'm doing something wrong.

**Comment by pzygielo** [comment]
I'm afraid the new, replacement [UnnecessaryImport](https://pmd.github.io/pmd-6.34.0/pmd_rules_java_codestyle.html#unnecessaryimport) rule reports the same FP.

**Comment by jonybuzz** [comment]
Still happening for this config:
```xml
...
    <properties>
        <java.version>11</java.version>
        <pmdVersion>6.41.0</pmdVersion>
    </properties>
    <build>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-pmd-plugin</artifactId>
                <version>3.15.0</version>
                <executions>
                    <execution>
                        <id>pmd-while-compiling</id>
                        <phase>validate</phase>
                        <goals>
                            <goal>check</goal>
                        </goals>
                        <configuration>
                            <printFailingErrors>true</printFailingErrors>
                        </configuration>
                    </execution>
                </executions>
                <dependencies>
                    <dependency>
                        <groupId>net.sourceforge.pmd</groupId>
                        <artifactId>pmd-core</artifactId>
                        <version>${pmdVersion}</version>
                    </dependency>
                    <dependency>
                        <groupId>net.sourceforge.pmd</groupId>
                        <artifactId>pmd-java</artifactId>
                        <version>${pmdVersion}</version>
                    </dependency>
                    <dependency>
                        <groupId>net.sourceforge.pmd</groupId>
                        <artifactId>pmd-javascript</artifactId>
                        <version>${pmdVersion}</version>
                    </dependency>
                    <dependency>
                        <groupId>net.sourceforge.pmd</groupId>
                        <artifactId>pmd-jsp</artifactId>
                        <version>${pmdVersion}</version>
                    </dependency>
                </dependencies>
            </plugin>
    </build>
...
```

**Comment by oowekyala** [comment]
@jonybuzz Can you please provide a code example? We have a passing regression test for this issue in our sources...

**Comment by lolo101** [comment]
Hi @jonybuzz this false positive still affects PMD 7.7.0

I have a code example that demonstrate this : https://github.com/lolo101/pmd-issue

**Comment by lolo101** [comment]
Opened new issue https://github.com/pmd/pmd/issues/5315
